scala/bug#10033 Remove items deprecated in 2.12 or earlier

NthPortal opened the pull request:
Removes most items deprecated in 2.12 and earlier, with the following
exceptions:

`scala.reflect.ClassManifestDeprecatedApis` is not removed, because it
is used by `scala.reflect.Manifest`, which is not deprecated. For the
same reason, the definitions of `ClassManifest` in the `scala.reflect`
package object are also not removed.

`scala.io.Position` is not removed, because it is is used by
`scala.io.Source.Positioner`, which is not deprecated. It may be
removed from the standard library with `scala.io.Source` (and perhaps
un-deprecated then).

Deprecations in `scala.collections` are skipped (unless referenced
outside the package), because the package is being rewritten for 2.13.

`scala.DelayedInit` is not removed, because there is not yet any
replacement implementation for `scala.App`.

`scala.runtime.StringAdd` is not removed, for bootstrap reasons. It can
be removed in 2.13.0-M5.

`scala.concurrent.Future.fold` and `scala.concurrent.Future.reduce` are
not removed, to facilitate cross-building between 2.11/2.12/2.13.

Fixes scala/bug#9641

Closes scala/bug#7308 (as won't fix)

NthPortal commented:
I've broken the changes up into different commits, per this comment: https://github.com/scala/bug/issues/10033#issuecomment-328679219. I'm happy to squash them if desired.

NthPortal commented:
I've also not marked this PR as fixing 10033 due to the `StringAdd` issue. I'll make a PR for 2.13.0-M5 to remove `StringAdd` and close the issue.

SethTisue commented:
this looks the appropriate granularity of commits to me.

adriaanm commented:
/rebuild

SethTisue commented:
/rebuild

NthPortal committed:
bug#10033 Remove deprecations from scala.{annotation, beans, runtime}

NthPortal committed:
bug#10033 Remove deprecations from scala.{sys, util, _}

NthPortal committed:
bug#10033 Remove deprecations from scala.text

Move scala.text.Document into compiler module
(in scala.tools.nsc.ast.TreeBrowser) and un-deprecate.

NthPortal committed:
bug#10033 Remove deprecations from scala.math

NthPortal committed:
bug#10033 Remove ClassManifest from Predef

Remove ClassManifest from Predef and associated places.

retronym commented:
Just to set expectations here: part of removing deprecated items is evaluating the effect on the community build. If a removal affects a small number of projects, we need to reach out to them and ask them to adapt (perhaps by sending a PR to the project). If causes widespread breakage, we might need to delay the change.

SethTisue commented:
good point, Jason. I'll try to be proactive about identifying PRs that need community build runs before merge (and about actually doing the runs, too).

NthPortal commented:
/sync

adriaanm commented:
/rebuild 3ba7918

adriaanm commented:
/rebuild d665bc815c244888ec615a32e41c10ec003d7ec7

NthPortal commented:
/sync

NthPortal commented:
Not sure how to get `combined` to update

NthPortal commented:
I'm not sure how community builds are run, but I've fixed up all the broken (par)tests at this point, so it's good to go I think.

hrhino committed:
Actually remove @bridge.

Follow-up to scala/scala#6268 / 1eb0fba.

Turns out that `rm src/library/scala/annotation/bridge.scala`
isn't the entire story: `Namers` references `bridge` by name,
and scaladoc uselessly checks `isBridge` (the compiler doesn't
set this flag until `erasure`, when the generated bridges show up).

Removed one of the scaladoc tests which tested that this annotation
didn't do more than it was supposed to.

NthPortal commented:
While looking into scala/bug#9554, I noticed that the information regarding how certain `Throwable`s are handled is present only in the docs for `onFailure`, and not `onComplete`. I'm going to copy the relevant info into the docs for `onComplete`, so it's not lost with these changes.

NthPortal commented:
/sync

NthPortal commented:
@SethTisue could you run a community build? Thanks!

SethTisue commented:
queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/951/ (link will be 404 til Jenkins queue clears)

NthPortal commented:
cheers!

SethTisue commented:
it's still running, but so far:

* scala-async is using `future` (lower case f), `Future.fold`, `Future.reduce`
* macro-paradise is using `tpnme.bridgeAnnot`
* scala-swing is using `Future#onSuccess`
* scalaz is using `PartialFunction.apply`

fixing these may uncover further errors in the same projects or in downstream projects

NthPortal commented:
So far, nothing looks terrible to fix

- scala-async is easy to fix - perhaps as easy as find-replace
- macro-paradise looks at first glance like it just needs the changes @hrhino made to be ported over; however, it's possible the relevant code is exposed to clients
- scala-swing with `Future.onSuccess` is annoying, because it's not a simple rename, but it still shouldn't be too hard (unless it's unclear what to do in the failure case)
- scalaz shouldn be reasonably easy to fix; it's not a simple rename, but the rewrite isn't too difficult

SethTisue commented:
remaining failures:

* genjavadoc — I think it's choking on the Scala version number, doesn't have to do with your change. I'll fix and re-run, since there are a number of downstream projects.
* scala-stm — uses `ClassManifest` (I left a comment on https://github.com/nbronson/scala-stm/issues/56)

SethTisue commented:
new run with genjavadoc fix: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/953/

SethTisue commented:
scalaz PR: https://github.com/scalaz/scalaz/pull/1654

SethTisue commented:
@viktorklang how do you feel about https://github.com/scala/scala/pull/6319/commits/b9e4b0b3eda89cfd3efffa01ff896bbf05825dca ? do you want the removals to go forward for 2.13?

NthPortal commented:
scala-async PR: scala/scala-async#188

I've discovered a problem, which is that `foldLeft` and `reduceLeft` didn't exist until 2.12, so it's difficult to remove `fold` and `reduce` from projects that support 2.11.

Should we hold off on removing `fold` and `reduce` until 2.14?

NthPortal commented:
scala-swing PR: scala/scala-swing#72 (builds fine)

NthPortal commented:
`ClassManifest` should be easy to fix as well because

- it's just an alias for `ClassTag` anyway, at this point
- it still exists in `scala.reflect`

I'm working on scala-java8-compat right now

NthPortal commented:
scala-java8-compat PR: scala/scala-java8-compat#100

viktorklang commented:
@SethTisue Thanks for the notice, Seth. I've commented.

SethTisue commented:
@NthPortal anything you add back in response to Viktor should also get a comment, I think

NthPortal commented:
@SethTisue do you want me to add it back in a separate commit, or rewrite b9e4b0b3eda89cfd3efffa01ff896bbf05825dca?

NthPortal commented:
@SethTisue @viktorklang what should be done about [`Future.fold` and `Future.reduce`](https://github.com/scala/scala/pull/6319#issuecomment-367558041) - leave them in for now?

NthPortal commented:
@viktorklang I left replies to some of your comments (and took care of the rest)

SethTisue commented:
> @SethTisue do you want me to add it back in a separate commit, or rewrite b9e4b0b?

rewrite, but it doesn't have to be right away

some folks prefer to squash/rebase as they go, other folks prefer to wait until all reviewers have signed off, then do it once at the very end. either way is fine.

viktorklang commented:
@NthPortal What is the problem with (removing) reduce and fold?

SethTisue commented:
@viktorklang the problem is that the alternatives, `reduceLeft` and `foldLeft`, weren't added until 2.12. so if someone wants to crossbuild against 2.11, 2.12, and 2.13, they'll have to find some kind of workaround.

this gets into a broader issue here of how we far we are willing to bend, in general, in order to preserve cross-building across more than two major Scala versions. cross-building against two major versions is definitely a goal of ours to support, against three it's much less clear that it's worth doing.

the argument for three over two is that there is a window — a window we are in _right now_ with 2.13 — where we want library authors to be supporting 2.13.0-Mx, but we definitely _don't_ expect them to drop 2.11 yet (not until 2.13.0 final is out!)

so, ugh, these decisions are kind of hard, actually. on something minor like this, perhaps the benefit of waiting for 2.14 to remove exceeds the cost of keeping the methods around (where they clutter up the code, clutter up Scaladoc, etc). idk

NthPortal commented:
The overhead cost for cross-building is non-trivial. It requires creating a utility object for each Scala version, with `foldLeft` or `reduceLeft` methods which forward to the appropriate method depending on Scala version. It's a decent amount of extra work and clutter, and the reason that I *didn't* just do that for scala-async to fix the PR.

viktorklang commented:
@SethTisue The code for reduce/fold is rather trivial, so any user who wants to maintain the exact same implementation should be able to do so. I don't have a good sense on how frequently used they are.

SethTisue commented:
@NthPortal what would just inlining the implementation look like, e.g. in scala-async?

NthPortal commented:
@SethTisue it probably looks like creating a utility object like this

```scala
private object FutureSpec {
  def foldLeft[T, R](futures: scala.collection.immutable.Iterable[Future[T]])(zero: R)(op: (R, T) => R)(implicit executor: ExecutionContext): Future[R] =
    foldNext(futures.iterator, zero, op)

  def reduceLeft[T, R >: T](futures: scala.collection.immutable.Iterable[Future[T]])(op: (R, T) => R)(implicit executor: ExecutionContext): Future[R] = {
    val i = futures.iterator
    if (!i.hasNext) Future.failed(new NoSuchElementException("reduceLeft attempted on empty collection"))
    else i.next() flatMap { v => foldNext(i, v, op) }
  }

  private[this] def foldNext[T, R](i: Iterator[Future[T]], prevValue: R, op: (R, T) => R)(implicit executor: ExecutionContext): Future[R] =
    if (!i.hasNext) Future.successful(prevValue)
    else i.next().flatMap { value => foldNext(i, op(prevValue, value), op) }
}
```

and then changing `Future.foldLeft => FutureSpec.foldLeft` and `Future.reduceLeft => FutureSpec.reduceLeft`

NthPortal commented:
Actually, I take that back. Having to worry about copyright significantly increases the cost of just inlining the current implementation. You need to worry about either including a copyright notice or changing the implementation so it's not a copy.

SethTisue commented:
yeah. I withdrew my inlining suggestion, but not until after you'd gotten the notification email, obviously :-/

NthPortal commented:
oops.

NthPortal committed:
bug#10033 Remove deprecations from scala.concurrent

NthPortal commented:
@SethTisue what do you think about `fold`/`reduce` at this point?

SethTisue commented:
> what do you think about fold/reduce at this point?

let's hold off on removing `foldLeft` and `reduceLeft` to 2.14. removing stuff is good, but we don't need to rush the timetable unless the old method is actually causing trouble

I'd suggest putting a comment on the methods, something like `// not removed yet in 2.13, to facilitate 2.11/2.12/2.13 cross-building; let's remove it in 2.14; see #6319` so the next boy/girl scout who comes through knows what's up

NthPortal commented:
Sounds good to me - that was my thought as well

NthPortal commented:
/sync

NthPortal commented:
/rebuild 691ffa7efc6fe4b3f2a1d1f5ef29564303560ffe
/rebuild ed614acf789b2ef5dd0624a90ea5f43714d5da0d

NthPortal commented:
/rebuild ed614acf789b2ef5dd0624a90ea5f43714d5da0d

adriaanm commented:
/synch

NthPortal commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -60,14 +60,14 @@ class FutureTests extends MinimalScalaTest {
 
       f2 foreach { _ => throw new ThrowableTest("dispatcher foreach") }
-      f2 onSuccess { case _ => throw new ThrowableTest("dispatcher receive") }
+      f2 onComplete { case Success(_) => throw new ThrowableTest("dispatcher receive"); case _ => }
```
@viktorklang You mentioned that `foreach` would make more sense than `onComplete` here. If that's the case, since the previous line already tests `foreach`, should this line just be removed? I don't know if the original intent was to check if `onSuccess` worked, and if it makes sense to test if `onComplete` works as well.

SethTisue replied:
@NthPortal I think you can safely assume Viktor was just reading through the diffs quickly and making style suggestions without realizing that this is in a special testing context

NthPortal replied:
ok

NthPortal commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -60,14 +60,14 @@ class FutureTests extends MinimalScalaTest {
 
       f2 foreach { _ => throw new ThrowableTest("current thread foreach") }
-      f2 onSuccess { case _ => throw new ThrowableTest("current thread receive") }
+      f2 onComplete { case Success(_) => throw new ThrowableTest("current thread receive"); case _ => }
```
@viktorklang Same question as [line 63](https://github.com/scala/scala/pull/6319/commits/691ffa7efc6fe4b3f2a1d1f5ef29564303560ffe#r171052827)

NthPortal commented:
/sync

SethTisue commented:
@NthPortal I think this might be ready for final review and merge, do you agree?

every downstream project doesn't need to be compatible with this in advance, we just want to assess whether it will be nbd to fix other projects up as needed, and it looks to me like we're fine there.

NthPortal commented:
@SethTisue sounds good to me

SethTisue commented:
@viktorklang I think this is good to go except for possibly https://github.com/scala/bug/issues/7308 ? that ticket appears not to have been noticed (since it was filed in 2013), I don't know if it has merit.

SethTisue commented:
🎉

SethTisue commented:
downstream:

* https://github.com/scala/scala-async/issues/191
* https://github.com/scalamacros/paradise/issues/114
* https://github.com/nbronson/scala-stm/issues/56

once the paradise one is addressed, that may uncover further issues farther downstream. (the other two are not pressing.)
